proposals/0030-core-image-story.md

```diff
@@ -0,0 +1,402 @@
+  attributes: Record<string, any>;
+}
+```
+
+> Specific types not necessarily accurate to implementation. A more or less open-ended type for `options` and `attributes` is nonetheless required, as different services are allowed to return different things (including non-standard HTML attributes.)
+
+## Image Services
+
+By default, Astro will ship with two services can users can choose to transform their images: Squoosh, and Sharp. Both services more or less offer the same results and mostly differ in where they can run. Squoosh will be the default, because albeit slower than Sharp, it supports more platforms.
+
+Keeping in line with how you can extend Astro in various ways (remark, rehype, integrations etc), it's possible for users to create their own services.
+
+Two types of services exists: Local and External.
+
+- Local services handle the image transformation directly at build in SSG / runtime in dev / SSR. You can think of those as wrapper around librairies like Sharp, ImageMagick or Squoosh.
+- External services point to URLs and can be used for adding support for services such as Cloudinary, Vercel or any RIAPI-compliant server.
+
+Services definitions take the shape of an exported default object with various methods ("hooks") used to create all the required properties. The major difference, API-wise, between Local and External services is the presence of a `transform` method doing the actual transformation.
+
+The different methods available are the following:
+
+**Required methods**
+
+- `getURL(options: ImageTransform): string`
+  - For local services, return the URL of the endpoint managing your transformation (in SSR and dev).
+  - For external services, return the final URL of the image.
+  - `options` contain the parameters passed by the user
+
+**Required for Local services only**
+
+- `parseURL(url: URL): ImageTransform`
+  - For SSR and dev, parses the generated URLs by `getURL` back into an ImageTransform to be used by `transform`.
+- `transform(buffer: Buffer, options: ImageTransform): { data: Buffer, format: OutputFormat }`
+  - Transform and return the image. It is necessary to return a `format` to ensure that the proper MIME type is served to users.
+
+Ultimately, it is up to the local endpoint (that `getURL` points to) to call both `parseURL` and `transform`. Those two methods are exposed as convention and used by the two base services, but your endpoint is free to achieve this in a different way if needed.
+
+**Optional**
+
+- `getHTMLAttributes(options: ImageTransform): Record<string, any>`
+  - Return all additional attributes needed to render the image in HTML. For instance, you might want to return a specific `class` or `style`, or `width` and `height`.
+
+Overall, it's important to remember that 99% of users won't create services, especially local ones. Adapters (think: Vercel's Image Optimization) and integrations (ex: A Cloudinary integration) will supply and configure services for the users.
```

Comment:
I love this callout. Can you share more about how users will customize this in practice? For example: 

- I'd assume Vercel's image service would be added by the Vercel adapter automatically, but only in production? Is that a correct assumption? And if so, how do I customize local (squoosh vs. sharp) during development.
- What if I want to use something like Cloudinary, is the new way to do that via an integration (`astro add cloudinary`)? If so, how does that conflict with Vercel if I also use the Vercel adapter? How would I pick either one?
- Did you give any thought to having built-in services, chosen via string? Something like `optimize: 'cloudinary'`? Or, integration adds the service, but then you still need to opt-into one specific service via an explicit string in config or via `Image` prop? 
- Speaking of, assuming switching between services on a per-image basis is out of scope? If so, then that would be worth adding to non-goals.

Reply:
> I love this callout. Can you share more about how users will customize this in practice? For example: 
> 
> - I'd assume Vercel's image service would be added by the Vercel adapter automatically, but only in production? Is that a correct assumption? And if so, how do I customize local (squoosh vs. sharp) during development.

Presumably you always want to use the same service in both dev and prod, otherwise you'll have different results? If you do want a different one however, you can do so the same way you can do different settings for dev and prod in the Astro config.

For Vercel's, I'd assume that it would work in dev when you run their CLI perhaps? I'm not sure..

> - What if I want to use something like Cloudinary, is the new way to do that via an integration (`astro add cloudinary`)? If so, how does that conflict with Vercel if I also use the Vercel adapter? How would I pick either one?

It would work the same way as other settings in Astro, which I must admit, I don't know. It's not necessarily intended that integrations set the service for you, but since they can change the Astro config, they can

User is always free to choose through the `image.service` property however

> - Did you give any thought to having built-in services, chosen via string? Something like `optimize: 'cloudinary'`? Or, integration adds the service, but then you still need to opt-into one specific service via an explicit string in config or via `Image` prop? 

I'm not sure I understand the question. This is kinda how it works.

`image.service` can point to any entrypoints, so you can do: 

- Install the `Cloudinary integration or package that has an entrypoint`
- Change `image.service` to `cloudinary-package-entrypoint`

Built in services, the two local ones, are also acceded this way, though they don't need to be installed.

> - Speaking of, assuming switching between services on a per-image basis is out of scope? If so, then that would be worth adding to non-goals.

Yep, out of scope because I'm not sure how to do it in SSR at the moment. Will add, thank you!

Reply:
> Built in services, the two local ones, are also acceded this way, though they don't need to be installed.

Ah good to know, I missed that above.

> Presumably you always want to use the same service in both dev and prod, otherwise you'll have different results? If you do want a different one however, you can do so the same way you can do different settings for dev and prod in the Astro config.

Well, sort of... the Vercel image optimization is only available in prod. It doesn't work locally since it only works by proxying images out of their own backend/storage. This is how Next.js does it: they use the local image optimizations (squoosh or sharp) in development, and then the Vercel magic `_vercel` endpoint in production.

Reply:
Updated with some information on how it's possible to set a different service using something like `process.env.MODE`